File: Data Structures & Algorithms/maximum-product-subarray/submission-5.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        result = minp = maxp = nums[0]

        for x in nums[1:]:
            candidates = (x, x*minp, x*maxp)
            maxp = max(candidates)
            minp = min(candidates)
            result = max(result, maxp)

        return result
        

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-5.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        int l = 0, r = nums.size()-1;

        while (l <= r) {
            if (nums[l] == target) {
                return l;
            }

            if (nums[r] == target) {
                return r;
            }
            
            int m = l + (r-l)/2;

            if (nums[m] == target) {
                return m;
            }

            if (target < nums[m]) {
                if (nums[m] < nums[l]) {
                    r = m-1;
                } else {
                    l = m+1;
                }
            } else {
                if (nums[m] < nums[l]) {
                    r = m-1;
                } else {
                    l = m+1;
                }
            }
        }

        return -1;
    }
};
